syntax = "proto3";

package agents;

option csharp_namespace = "Microsoft.AutoGen.Contracts";

import "cloudevent.proto";
import "google/protobuf/any.proto";

message AgentId {
    string type = 1;
    string key = 2;
}

message RegisterAgentTypeRequest {
    string request_id = 1;
    string type = 2;
}

message RegisterAgentTypeResponse {
    string request_id = 1;
    bool success = 2;
    optional string error = 3;
}

message TypeSubscription {
    string topic_type = 1;
    string agent_type = 2;
}

message TypePrefixSubscription {
    string topic_type_prefix = 1;
    string agent_type = 2;
}

message Subscription {
    oneof subscription {
        TypeSubscription typeSubscription = 1;
        TypePrefixSubscription typePrefixSubscription = 2;
    }
}

message AddSubscriptionRequest {
    string request_id = 1;
    Subscription subscription = 2;
}

message AddSubscriptionResponse {
    string request_id = 1;
    bool success = 2;
    optional string error = 3;
}

service AgentRpc {
    rpc OpenChannel (stream Message) returns (stream Message);
    rpc GetState(AgentId) returns (GetStateResponse);
    rpc SaveState(AgentState) returns (SaveStateResponse);
}

message AgentState {
  AgentId agent_id = 1;
  string eTag = 2;
  oneof  data {
    bytes binary_data = 3;
    string text_data = 4;
    google.protobuf.Any proto_data = 5;
  }
}

message GetStateResponse {
	AgentState agent_state = 1;
	bool success = 2;
    optional string error = 3;
}

message SaveStateResponse {
	bool success = 1;
    optional string error = 2;
}

message Message {
    oneof message {
        io.cloudevents.v1.CloudEvent cloudEvent = 1;
        RegisterAgentTypeRequest registerAgentTypeRequest = 2;
        RegisterAgentTypeResponse registerAgentTypeResponse = 3;
        AddSubscriptionRequest addSubscriptionRequest = 4;
        AddSubscriptionResponse addSubscriptionResponse = 5;
    }
}

